### PR TITLE
Prefer to set task limits from requests when only the latter is set

### PR DESCRIPTION
**Pull request by katrogan**
# TL;DR
Users report that 

> we're noticing that setting resource requests, but not limits, for tasks is causing the requests to not be honored. it appears that the platform defaults are getting set for the limits, and copied into requests. is this intended?

The source of this is that an earlier change erroneously preferred to assign unset limits from the admin application config rather than set request values (when they were set).

Tested on flytesandbox

## Type
 - [x] Bug Fix
 - [ ] Feature
 - [ ] Plugin

## Are all requirements met?

 - [x] Code completed
 - [x] Smoke tested
 - [x] Unit tests added
 - [x] Code documentation added
 - [ ] Any pending items have an associated Issue

## Complete description
 _How did you fix the bug, make the feature etc. Link to any design docs etc_

## Tracking Issue
https://github.com/flyteorg/flyte/issues/1450

## Follow-up issue
_NA_

### CODE REVIEW

File: pkg/manager/impl/execution_manager.go [hunk 1]
```diff
@@ -234,26 +234,36 @@ func createTaskDefaultLimits(ctx context.Context, task *core.CompiledTask,
 	}
 	taskResourceLimits := runtimeInterfaces.TaskResourceSet{}
 
-	// For resource values, we prefer to use the limits set in the application config over the set resource values.
-	if !configResourceLimits.CPU.IsZero() {
+	// For resource values, we prefer to set unset limits to the request values over the config values.
+	if cpuIndex >= 0 {
+		cpuLimit = resourceRequestEntries[cpuIndex].Value
```
Comment (kumare3) [inline pkg/manager/impl/execution_manager.go:239]: why do we parse this multiple times? Why not parse it once at load time and then use? this will cause failuers much later, incase of a bug in the config?

File: pkg/manager/impl/execution_manager.go [hunk 2]
```diff
@@ -234,26 +234,36 @@ func createTaskDefaultLimits(ctx context.Context, task *core.CompiledTask,
...
+		cpu, err := resource.ParseQuantity(cpuLimit)
+		if err != nil {
+			logger.Errorf(ctx, "Failed to parse user cpu limit from task spec [%s] with err [%+v]", cpuLimit, err)
```
Comment (kumare3) [inline pkg/manager/impl/execution_manager.go:242]:
```suggestion
			logger.Errorf(ctx, "In Task [%s] Failed to parse specified CPU limits [%s]. Resources should be valid Kubernetes quantities. Reason [%s]", task.Name, cpuLimit, err.String())
```

maybe `task.Name` is incorrect, but please provide the actual name of the task. Errors are extremely important. Obtuse errors just increase confusion

File: pkg/manager/impl/execution_manager.go [hunk 3]
```diff
@@ -234,26 +234,36 @@ func createTaskDefaultLimits(ctx context.Context, task *core.CompiledTask,
...
 		} else {
 			taskResourceLimits.CPU = cpu
 		}
 	}
-	if !configResourceLimits.Memory.IsZero() {
+	if memoryIndex >= 0 {
+		memoryLimit = resourceRequestEntries[memoryIndex].Value
+		memory, err := resource.ParseQuantity(memoryLimit)
+		if err != nil {
+			logger.Errorf(ctx, "Failed to parse user memory limit from task spec [%s] with err [%+v]", memoryLimit, err)
```
Comment (kumare3) [inline pkg/manager/impl/execution_manager.go:260]: same as above, better errors please

File: pkg/manager/impl/execution_manager.go [hunk 4]
```diff
@@ -234,26 +234,36 @@ func createTaskDefaultLimits(ctx context.Context, task *core.CompiledTask,
...
+		}
+	} else if !configResourceLimits.Memory.IsZero() {
 		taskResourceLimits.Memory = configResourceLimits.Memory
 	} else {
-		if memoryIndex >= 0 {
-			memoryLimit = resourceRequestEntries[memoryIndex].Value
-		}
 		memory, err := resource.ParseQuantity(memoryLimit)
 		if err != nil {
 			logger.Errorf(ctx, "Failed to parse user memory limit from task spec [%s] with err [%+v]", memoryLimit, err)
```
Comment (kumare3) [inline pkg/manager/impl/execution_manager.go:269]: same as above

File: pkg/manager/impl/execution_manager.go [hunk 5]
```diff
@@ -265,18 +275,19 @@ func createTaskDefaultLimits(ctx context.Context, task *core.CompiledTask,
...
 		taskResourceLimits.GPU = configResourceLimits.GPU
 	}
-	if !configResourceLimits.EphemeralStorage.IsZero() {
-		taskResourceLimits.EphemeralStorage = configResourceLimits.EphemeralStorage
-	} else if ephemeralStorageIndex >= 0 {
+
+	if ephemeralStorageIndex >= 0 {
 		ephemeralStorage, err := resource.ParseQuantity(resourceRequestEntries[ephemeralStorageIndex].Value)
 		if err != nil {
 			logger.Errorf(ctx, "Failed to parse user ephemeral storage limit from task spec [%s] with err [%+v]",
```
Comment (kumare3) [inline pkg/manager/impl/execution_manager.go:282]: same as above

File: pkg/manager/impl/execution_manager.go [hunk 6]
```diff
@@ -234,26 +234,36 @@ func createTaskDefaultLimits(ctx context.Context, task *core.CompiledTask,
...
+		}
+	} else if !configResourceLimits.CPU.IsZero() {
 		taskResourceLimits.CPU = configResourceLimits.CPU
 	} else {
-		if cpuIndex >= 0 {
-			cpuLimit = resourceRequestEntries[cpuIndex].Value
-		}
 		cpu, err := resource.ParseQuantity(cpuLimit)
 		if err != nil {
 			logger.Errorf(ctx, "Failed to parse user cpu limit from task spec [%s] with err [%+v]", cpuLimit, err)
```
Comment (pmahindrakar-oss) [inline pkg/manager/impl/execution_manager.go:251]: Wont this value be `cpuLimit := "500m"` and if yes do we need to handle the parsing and do error handling . If its const then we can probably have a right quantity set . 

Also the error message might need to change if thats the case

File: pkg/manager/impl/execution_manager.go [hunk 7]
```diff
@@ -234,26 +234,36 @@ func createTaskDefaultLimits(ctx context.Context, task *core.CompiledTask,
...
+		} else {
+			taskResourceLimits.Memory = memory
+		}
+	} else if !configResourceLimits.Memory.IsZero() {
 		taskResourceLimits.Memory = configResourceLimits.Memory
 	} else {
-		if memoryIndex >= 0 {
-			memoryLimit = resourceRequestEntries[memoryIndex].Value
-		}
 		memory, err := resource.ParseQuantity(memoryLimit)
```
Comment (pmahindrakar-oss) [inline pkg/manager/impl/execution_manager.go:267]: Same here .
`memoryLimit := "500Mi"`

File: pkg/manager/impl/execution_manager.go (hunk 6 above)
Comment (katrogan) [reply pkg/manager/impl/execution_manager.go]: talked with Haytham - since we never expect to use these defaults (represents a configuration error if so) we'll log an error when the task limits end up resolving to unset.

File: pkg/manager/impl/execution_manager.go (hunk 7 above)
Comment (katrogan) [reply pkg/manager/impl/execution_manager.go]: see above

File: pkg/manager/impl/execution_manager.go (hunk 2 above)
Comment (katrogan) [reply pkg/manager/impl/execution_manager.go]: updated

File: pkg/manager/impl/execution_manager.go (hunk 3 above)
Comment (katrogan) [reply pkg/manager/impl/execution_manager.go]: code has been refactored but the full task id is in the error msg

File: pkg/manager/impl/execution_manager.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (katrogan) [reply pkg/manager/impl/execution_manager.go]: yes, the index is from `resourceRequestEntries`
